src/nosimg-enc.c

```diff
@@ -0,0 +1,127 @@
+// SPDX-License-Identifier: GPL-2.0-or-later
+/*
+ * nosimg-enc.c - encode/decode "nos.img" image for XikeStor SKS8300 series
+ */
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <stdbool.h>
+#include <stdint.h>
+#include <unistd.h>
+
+#define ENCODE_BLKLEN	0x100
+#define ENCODE_BLOCKS	2
+
+static uint32_t patterns[] = {
+	0xeeddcc21,  0x5355eecc,  0xdd55807e,  0x00000000,
+	0xcdbddfae,  0xbb9b8901,  0x70e5ccdd,  0xf6fc8364,
+	0xecddcef1,  0xe354fed0,  0xbdabdde1,  0xe4b4d583,
+	0xedfed0cd,  0xb655cca3,  0xedd5c67e,  0xddcc2153,
+	0xec4ddc00,  0x5355cdc3,  0x2201807e,  0xefbc7566,
+	0xa6c0cc2f,  0xfed0eecc,  0xdd550101,  0x0101c564,
+	0x9945ab32,  0x55807eef,  0x55807eef,  0xbc756689,
+	0xe31d83dd,  0xfe558eab,  0x7d55807e,  0xff01ac66,
+	0x0ec992d9,  0x73e50101,  0xbde510ce,  0x0101bae8,
+	0x3edd81a1,  0x53330101,  0x9ac510aa,  0x01ce8ae1,
+	0xb1fb0080,  0x53770000,  0x70dc0001,  0x0000cbb1,
+	0xa0300000,  0x55a60000,  0xcabd0101,  0x0000c9b2,
+	0x81900100,  0x5a210001,  0x79bc0100,  0x78007bb3,
+	0xd4970100,  0x5355a9fc,  0xdda501be,  0xafc175c5,
+	0x8ed77700,  0x55d00dac,  0x0155807e,  0xefbc7ee6,
+	0xf16c5200,  0x331698cc,  0x01010101,  0x00007988,
+};
+
+static void __attribute__((noreturn)) usage(void)
+{
+	fprintf(stderr, "Usage: nosimg-enc -i infile -o outfile [-d]\n");
+	exit(EXIT_FAILURE);
+}
+
+static void EncodeBlock100(uint8_t *data, bool decode)
+{
+	int i;
+
+	for (i = 0; i < ENCODE_BLKLEN; i++)
+	{
+		uint32_t tmp;
+		tmp = patterns[i / sizeof(uint32_t)];
+		tmp >>= (sizeof(uint32_t) - i % sizeof(uint32_t) - 1) * 8;
+		data[i] -= (tmp & 0xff) * (decode ? -1 : 1);
+	}
+}
+
+int main(int argc, char **argv)
+{
+	int i, c, n, ret = EXIT_SUCCESS;
+	char *ifn = NULL, *ofn = NULL;
+	bool decode = false;
+	uint8_t buf[0x10000];
```

Comment:
Does this need to be bigger than `BLKLEN`?

Reply:
It's for copying remaining data. (only the first 512 bytes (`0x200`) are encoded in this format)

- `buf[0x100]`

  ```
  $ time ./build/nosimg-enc -i test.dec -o test.bin

  real    0m5.133s
  user    0m0.023s
  sys     0m0.171s
  ```

- `buf[0x10000]`

  ```
  $ time ./build/nosimg-enc -i test.dec -o test.bin

  real    0m0.143s
  user    0m0.006s
  sys     0m0.000s
  ```

Reply:
Aha, I somehow overlooked the second loop. Isn't 64k a bit big to put on the stack?

Reply:
Hm, I reduced the buffer to 4k (`0x1000`).